### PR TITLE
feat(airbyte-cdk): Add extra fields to StreamSlice

### PR DESCRIPTION
**Pull request by tolik0**
## What

This PR introduces the ability to include extra fields from parent records into the StreamSlice. This feature allows for additional data, beyond the partition key, to be extracted and utilized in downstream processes.
Resolves: https://github.com/airbytehq/airbyte-internal-issues/issues/9909 

## How

- Added support for defining extra fields in the `ParentStreamConfig`.
- The extra fields are specified as paths and are extracted from the parent record during the `StreamSlice` generation.
- Enhanced the `SubstreamPartitionRouter` to include these extracted fields within each StreamSlice.

## User Impact

- Users can now define additional field paths in their configuration that will be included in each StreamSlice.
- This offers more flexibility in how data is extracted and used for further processing within streams.

## Can this PR be safely reverted and rolled back?

- [ ] YES 💚
- [ ] NO ❌

### CODE REVIEW

**Comment by tolik0** [comment]
Regression test result: https://github.com/airbytehq/airbyte/actions/runs/11181168643/job/31084544603

**Review by bazarnov — APPROVED** [review]

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/extractors/record_filter.py [hunk 1]
```diff
@@ -33,7 +33,12 @@ def filter_records(
         stream_slice: Optional[StreamSlice] = None,
         next_page_token: Optional[Mapping[str, Any]] = None,
     ) -> Iterable[Mapping[str, Any]]:
-        kwargs = {"stream_state": stream_state, "stream_slice": stream_slice, "next_page_token": next_page_token}
+        kwargs = {
+            "stream_state": stream_state,
+            "stream_slice": stream_slice,
+            "next_page_token": next_page_token,
+            "stream_slice.extra_fields": stream_slice.extra_fields if stream_slice else {},
+        }
```
Comment (brianjlai) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/extractors/record_filter.py:41]: can we add a test here that validates if we call `filter_records()` on the nested `stream_slice.extra_fields` context that we get the expected results

I only did a quick check so i might not have set up the interpolation right, but when I added a small test using `"{{ stream_slice.extra_fields.some_key }}"` and modified kwargs to look like this:
```
kwargs = {
            "stream_state": stream_state,
            "stream_slice": stream_slice,
            "next_page_token": next_page_token,
            "stream_slice.extra_fields": {"some_key": "some_value", "some_bool": True},
        }
```

I wasn't able to get this to resolve during the interpolation step to a real slice value. Once we add a unit test to confirm, then happy to 👍 

Do we need to nest the entire `extra_field` object under the `stream_slice` key in `kwargs`?

**Review by lazebnyi — APPROVED** [review]

**Review by brianjlai — COMMENTED** [review]
mainly i have a question around how we define `stream_slice.extra_fields` in the kwargs so that it's available during interpolation. Otherwise in terms of how the code is structures looks good

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/extractors/record_filter.py (hunk 1 above)
Comment (tolik0) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/extractors/record_filter.py:41]: The value can be accessed using `stream_slice.extra_fields["some_key"]`. I've added a test to validate this behaviour, and it has been tested during the regression tests for Jira.

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_component_schema.yaml [hunk 2]
```diff
@@ -2080,6 +2080,18 @@ definitions:
         description: Indicates whether the parent stream should be read incrementally based on updates in the child stream.
         type: boolean
         default: false
+      extra_fields:
+        title: Extra Fields
+        description: Array of field paths to include as additional fields.
```
[the remark this answers is not in the document]
Comment (tolik0) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_component_schema.yaml]: Updated the description

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_component_schema.yaml [hunk 3]
```diff
@@ -2080,6 +2080,18 @@ definitions:
         description: Indicates whether the parent stream should be read incrementally based on updates in the child stream.
         type: boolean
         default: false
+      extra_fields:
```
[the remark this answers is not in the document]
Comment (tolik0) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_component_schema.yaml:2083]: Added the `interpolation_context`

**Review by brianjlai — APPROVED** [review]
lgtm!

**Pull request merged**